source/opt/instruction.h

```diff
@@ -506,28 +512,27 @@ inline void Instruction::SetInOperands(OperandList&& new_operands) {
 }
 
 inline void Instruction::SetResultId(uint32_t res_id) {
-  result_id_ = res_id;
-  auto ridx = (type_id_ != 0) ? 1 : 0;
+  auto ridx = has_type_id_ ? 1 : 0;
```

Comment:
There's a weirdness here.  If res_id is 0, should we also update the has_result_id_ flag?  And vice versa?

I think we should. Needs
` has_result_id_ = res_id !=0;`

Reply:
Or, maybe it should be assert(res_id != 0); If you  have a result_id and you set it to 0 here you'll still have an entry in the operand array for it which would be bad. I think we want to just disallow that from happening?

Reply:
I think we should assert.  Before this change if we did this change we would of had an error when trying to remove the result type (see below).  We can rely on `ToNop` to create a Nop.

If we need to be able to delete a result from an instruction because we changed it to something else, we can see the use case.

Reply:
Done.